ouroboros-network/test-cddl/Main.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+module Main
+where
+
+import System.Environment (getArgs)
+import System.Exit (ExitCode(..))
+import System.Process.ByteString.Lazy
+import Control.Monad
+import Control.Exception.Base (throw)
+import qualified Data.ByteString.Lazy as BS
+import Data.ByteString.Lazy (ByteString)
+import Data.ByteString.Lazy.Char8 as Char8 (putStrLn, unpack, lines)
+import qualified Codec.Serialise.Class as Serialise
+import Codec.CBOR.Decoding (decodeWord, decodeListLenOf, decodeBytes)
+import Codec.CBOR.Read
+
+import Ouroboros.Network.Protocol.ChainSync.Type as CS
+import Ouroboros.Network.Protocol.ChainSync.Codec (codecChainSync)
+import Network.TypedProtocol.ReqResp.Type as ReqResp
+import Ouroboros.Network.Protocol.ReqResp.Codec (codecReqResp)
+import Ouroboros.Network.Protocol.PingPong.Codec (codecPingPong)
+import Network.TypedProtocol.PingPong.Type as PingPong
+import Ouroboros.Network.Protocol.BlockFetch.Codec (codecBlockFetch)
+import Ouroboros.Network.Protocol.BlockFetch.Type as BlockFetch
+import Ouroboros.Network.Testing.ConcreteBlock
+
+import Network.TypedProtocol.Codec
+
+-- the concrete types used for the test
+type CS = ChainSync DummyBytes DummyBytes
+type RR = ReqResp DummyBytes DummyBytes
+type BF = BlockFetch BlockHeader BlockBody
+
+main :: IO ()
+main = getArgs >>= \case
+    [cddlCmd,diag2cborCmd,cddlSpec,rounds]
+        -> generateAndDecode cddlCmd diag2cborCmd cddlSpec (read rounds)
+    _ -> error "call with 4 arguments"
+
+test :: IO ()
+test = generateAndDecode "./cddl" "./diag2cbor.rb" "../doc/messages.cddl" 100
```

Comment:
If we make this test be what happens when `main` is called with no args then we can turn this exe into a cabal test-suite.